### DISCUSSION TITLE
Unable to find application named 'Simulator'?

### DISCUSSION THREAD

**Discussion by Happypig375** [community/community#25331] [Actions]
<p>This is the run:<br>
</p><aside class="onebox whitelistedgeneric">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/verybadcat/CSharpMath/runs/776548293?check_suite_focus=true" target="_blank" rel="nofollow noopener">GitHub</a>
  </header>
  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/19922066/181090459-ad616147-f43f-4037-8684-875266ead7b2.png" class="thumbnail onebox-avatar" width="420" height="420">

<h3><a href="https://github.com/verybadcat/CSharpMath/runs/776548293?check_suite_focus=true" target="_blank" rel="nofollow noopener">verybadcat/CSharpMath</a></h3>

<p>LaTeX. in C#. (ported from the wonderful iosMath project). - verybadcat/CSharpMath</p>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>The line</p>
<pre><code class="lang-auto">open -a 'Simulator' --args -CurrentDeviceUDID $IOS_SIM_UDID
</code></pre>
<p>seems to output</p>
<pre><code class="lang-auto">Unable to find application named 'Simulator'
</code></pre>
<p>.</p>
<p>Is there any way to start the iOS Simulator? It is clearly listed in <a href="https://github.com/actions/virtual-environments/blob/40caaf3e2c11a9e4289482cce72e14ed97fa64fe/images/macos/macos-10.15-Readme.md#installed-simulators" rel="nofollow noopener">https://github.com/actions/virtual-environments/blob/40caaf3e2c11a9e4289482cce72e14ed97fa64fe/images/macos/macos-10.15-Readme.md#installed-simulators</a> but I cannot find any way to use it.</p>

**Comment by Yanjingzhu** [comment]
<p>Hi <a class="mention" href="/happypig375">@happypig375</a>,</p>
<p>According to the error message:<br>
<code>Unable to find application named 'Simulator'</code><br>
I guess there is some settings missing for Simulator which points to the application path.</p>
<p>I run <code>xcrun simctl list</code> , the ID for iPhone 11 device iOS 13.5 is<br>
<code>67864C3A-1C15-48B9-96E5-EA9B279F2253</code>.<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/9/9b03b8d080a875e04d8c68d1d30605fc925cbc95.png" data-download-href="/uploads/short-url/m7jYTgWKxMVDJjoB0p9h62r5rZr.png?dl=1" title="iphone 11" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181090461-07026854-c5e1-4910-9ff5-323a1ee6313a.png" alt="iphone 11" data-base62-sha1="m7jYTgWKxMVDJjoB0p9h62r5rZr" width="690" height="453" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/9/9b03b8d080a875e04d8c68d1d30605fc925cbc95_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">iphone 11</span><span class="informations">748×492 20.3 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>To start the iOS simulator, you could try</p>
<pre><code class="lang-auto">xcrun simctl boot $IOS_SIM_UDID
</code></pre>

**Comment by Happypig375** [comment]
<aside class="quote group-github-partners" data-username="yanjingzhu" data-post="2" data-topic="118331">
<div class="title">
<div class="quote-controls"></div>
 yanjingzhu:</div>
<blockquote>
<p>To start the iOS simulator, you could try</p>
<pre><code class="lang-auto">xcrun simctl boot $IOS_SIM_UDID
</code></pre>
</blockquote>
</aside>
<p>According to <a href="https://stackoverflow.com/a/35821607/5429648" rel="nofollow noopener">https://stackoverflow.com/a/35821607/5429648</a><br>
, that is not a full boot and some services (including those provided by the system app, like install or launch) are unavailable in this mode.</p>

**Comment by Happypig375** [comment]
<p>New run with correct UDID, error still persists. <a href="https://github.com/verybadcat/CSharpMath/runs/779554691?check_suite_focus=true" rel="nofollow noopener">https://github.com/verybadcat/CSharpMath/runs/779554691?check_suite_focus=true</a></p>

**Comment by Happypig375** [comment]
<p><a class="mention" href="/yanjingzhu">@yanjingzhu</a> Is there any way to solve this?</p>

**Comment by Yanjingzhu** [comment, marked as answer]
<p>I was trying hard to find a way to resolve your issue. But I met another issue .</p>
<p>Firstly, I found that the UID has changed today. There is a method to get uid of a simulator:</p>
<pre><code class="lang-auto">get_uid() {
  echo $( xcrun simctl list | grep -w $SIMULATOR_NAME | awk 'match($0, /\(([-0-9A-F]+)\)/) { print substr( $0, RSTART + 1, RLENGTH - 2 )}' )
}
</code></pre>
<p>And I tried to change Simulator to the exact path</p>
<pre><code class="lang-auto">xcrun simctl boot $IOS_SIM_UDID
open -a '/Applications/Xcode.app/Contents/Developer/Applications/Simulator.app/Contents/MacOS/Simulator' --args -CurrentDeviceUDID $IOS_SIM_UDID
</code></pre>
<p>The scripts stuck at</p>
<pre><code class="lang-auto">   echo "Installing app at path $APP_PATH"
   xcrun simctl install booted $APP_PATH
</code></pre>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/e/ed1f2d8625ce6a58d86fff699f1bd1eb66a06000.png" data-download-href="/uploads/short-url/xPFXrbx8b4P6LI1h1tzTwBtGkOQ.png?dl=1" title="workflow stuck" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181090463-96248d89-0e7a-49fa-96ca-551465b30c60.png" alt="workflow stuck" data-base62-sha1="xPFXrbx8b4P6LI1h1tzTwBtGkOQ" width="499" height="499" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/e/ed1f2d8625ce6a58d86fff699f1bd1eb66a06000_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">workflow stuck</span><span class="informations">804×804 20.1 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div><br>
<div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/f/ff395cf31e78b464230416614210e166b4db4874.png" data-download-href="/uploads/short-url/ApOEB4Lr7cYAj5DcEcRDzlP7wi0.png?dl=1" title="install app" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181090464-268564de-582f-45f7-b794-5d2fbb8ce532.png" alt="install app" data-base62-sha1="ApOEB4Lr7cYAj5DcEcRDzlP7wi0" width="690" height="378" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/f/ff395cf31e78b464230416614210e166b4db4874_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">install app</span><span class="informations">774×425 21.8 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>As I am not familiar with ios, could you please check whether this script could run successfully in your local machine?</p>

**Comment by Happypig375** [comment]
<p>Thank you! :heart: It now matches my local output.<br>
The local script is:</p>
<pre><code class="lang-auto">

#!/bin/bash
        nuget restore CSharpMath.sln
        msbuild CSharpMath.Ios.Tests
        export IOS_SIM_NAME='iPhone 11' # https://github.com/actions/virtual-environments/blob/fab29f60777e2532bfaa6578efdaca0f6dc05210/images/macos/macos-10.15-Readme.md#installed-simulators
        export BUNDLE_IDENTIFIER=CSharpMath.Ios.Unit-Tests # Located inside CSharpMath.Ios.Tests Info.plist file
        export APP_PATH=CSharpMath.Ios.Tests/bin/iPhoneSimulator/Debug/CSharpMath.Ios.Tests.app


        # run_ios_sim builds and runs an iOS app on the simulator
        #
        # It is designed to replicate the behavior of "Run" in Xcode and assumes basic
        # xcodebuild usage.
        # 
        # USAGE:
        # export IOS_SIM_UDID=342F9A20-DF48-41A9-BE60-C6B35F47E97F; \
        # export BUNDLE_IDENTIFIER=a.Some; \
        # export APP_PATH=$PWD/Build/Debug-iphonesimulator/$APP_NAME.app \
        # /path/to/run_ios_sim.sh 
        #
        # Note that the UDID must match a device where runtime is installed See
        # available devices with "simctl list"
        #
        # Additionally, it supports the ability to wait and attach LLDB in a similar
        # fashion that Xcode does.
        # export DEBUGGER_ENABLED=1|0
        #
        # Tested on Xcode 8.3.1 a few times
        # Author: Jerry Marino - @jerrymarino

        APP_NAME=`echo ""${APP_PATH##*/}"" | cut -d'.' -f1`

        echo "Running sim for $APP_NAME - $BUNDLE_IDENTIFIER"

        IOS_SIM_UDID=`xcrun simctl list | grep -w $SIMULATOR_NAME | awk 'match($0, /\(([-0-9A-F]+)\)/) { print substr( $0, RSTART + 1, RLENGTH - 2 )}'`

        SIMULATOR_PATH='/Applications/Xcode.app/Contents/Developer/Applications/Simulator.app/Contents/MacOS/Simulator'


        # If the booted simulator does not match, then we need to restart it. Expect an
        # output list of the form
        # "Phone: iPhone 7 Plus (342F9A20-DF48-41A9-BE60-C6B35F47E97F) (Booted)"
        BOOTED_UDID=`xcrun simctl list | grep Booted | perl -pe 's/(.*\()(.*)\)+ (.*)/\2/' | sed -n 1p`
        if [[ $BOOTED_UDID != $IOS_SIM_UDID ]]; then
          killall $SIMULATOR_PATH
        else
          # FIXME: We don't need to do this for all cases and
          # it is slow
          killall $SIMULATOR_PATH
        fi

        # Open the simulator
        open -a $SIMULATOR_PATH --args -CurrentDeviceUDID $IOS_SIM_UDID

        # Wait until there is a device booted

        function booted_sim_ct() {
          echo `xcrun simctl list | grep Booted | wc -l | sed -e 's/ //g'`
        }

        while [ `booted_sim_ct` -lt 1 ]
        do
          sleep 1
        done

        echo "Installing app at path $APP_PATH"
        xcrun simctl install booted $APP_PATH

        if [[ $DEBUGGER_ENABLED == "1" ]]; then
          LAUNCH_DEBUGGER_ENABLED_FLAG=--wait-for-debugger
          USE_CONSOLE_FLAG=""
        else
          USE_CONSOLE_FLAG=--console
          LAUNCH_DEBUGGER_ENABLED_FLAG=""
        fi

        LOG_FILE=/tmp/run_ios_sim.log
        echo "Starting Sim for $APP_PATH" &gt; $LOG_FILE

        # Launch the app program into the booted sim
        # - Pipe the output to a log file
        # - Run in the background
        `xcrun simctl launch $LAUNCH_DEBUGGER_ENABLED_FLAG $USE_CONSOLE_FLAG booted $BUNDLE_IDENTIFIER 2&gt;&amp;1 &gt;&gt; $LOG_FILE` &amp;

        LAUNCH_SHELL_PID=$!

        function newest_related_process() {
          echo `ps aux | grep $APP_NAME.app/$APP_NAME | grep -v -e "*.grep" | sort -k 4nr,4  | awk '{ print ( $2 &gt; '$LAUNCH_SHELL_PID' ) ? $2 : "" }' | grep -v -e '^$' | head -1`
        }

        # Wait for the process to launch. We need to get the PID of the launched App
        _Z=`newest_related_process`
        _X=0
        while [[ "$_Z" != "$_X" ]]
        do
          sleep 1
          _Z=$_X
          _X=`newest_related_process`
          echo "waiting for process.."
        done

        APP_PID=$_X

        if [[ $DEBUGGER_ENABLED == "1" ]]; then
          # Manually startup LLDB.
          echo "$APP_NAME running at pid $APP_PID, waiting for lldb.."
          lldb -p $APP_PID
        else
          # If not debugging read the log file
          # when the log closes, the app will terminate
          echo "$APP_NAME running at pid $APP_PID"
          tail -f $LOG_FILE
        fi</code></pre>

**Comment by Happypig375** [comment]
<p>On local testing, the app has successfully launched but the script seems to not able to detect it. Some more Googling will be required.</p>

**Comment by Happypig375** [comment]
<p>If I add add two <code>echo</code>es to the problematic loop,</p>
<pre><code class="lang-auto">        while [[ "$_Z" != "$_X" ]]
        do
          sleep 1
          _Z=$_X
          _X=`newest_related_process`
          echo Z is $_Z
          echo X is $_X
          echo "waiting for process.."
        done
</code></pre>
<p>, the output becomes</p>
<pre><code class="lang-auto">Z is 0
X is 2301
waiting for process..
Z is 2301
X is 2311
waiting for process..
Z is 2311
X is 2321
waiting for process..
Z is 2321
X is 2331
waiting for process..
Z is 2331
X is 2341
waiting for process..
Z is 2341
X is 2351
waiting for process..
Z is 2351
X is 2361
waiting for process..
Z is 2361
X is 2371
waiting for process..
Z is 2371
X is 2381
waiting for process..
Z is 2381
X is 2391
waiting for process..
</code></pre>

**Comment by Happypig375** [comment]
<p>I have replaced</p>
<pre><code class="lang-auto">ps aux | grep $APP_NAME.app/$APP_NAME | grep -v -e "*.grep" | sort -k 4nr,4  | awk '{ print ( $2 &gt; '$LAUNCH_SHELL_PID' ) ? $2 : "" }' | grep -v -e '^$' | head -1
</code></pre>
<p>with</p>
<pre><code class="lang-auto">pgrep $APP_NAME -P $LAUNCH_SHELL_PID | head -1
</code></pre>
<p>and it seems to work correctly.</p>

**Comment by Yanjingzhu** [comment]
<p>I am very glad to hear that you have resolved your issue. If you meet any other issue when using GitHub Actions, welcome to our community again. :smiley:</p>

**Comment by Happypig375** [comment]
<p>Thanks <a class="mention" href="/yanjingzhu">@yanjingzhu</a>! The final form of the workflow file can be seen in <a href="https://github.com/verybadcat/CSharpMath/pull/133/files#diff-bb44396ad1769acfefe20015d1f4a095" rel="nofollow noopener">https://github.com/verybadcat/CSharpMath/pull/133/files#diff-bb44396ad1769acfefe20015d1f4a095</a> where an app file is loaded into the simulator and the app logs are extracted for checking for test failures.</p>
